### PR TITLE
Fix wxWindow::Refresh not generating paint events

### PR DESCRIPTION
**Pull request by ffa-grahamdawes**
Under wxQT wxWindow::Refresh was not generating paint events.  This has now been fixed.

### CODE REVIEW

File: src/qt/window.cpp [hunk 1]
```diff
@@ -393,13 +393,35 @@ void wxWindowQt::WarpPointer(int x, int y)
...
+
+    wxWindowDC dc(this);
+
+    const int id = GetId();
+
+    wxEraseEvent eraseEvent( id, &dc );
+    eraseEvent.SetEventObject( this );
+    HandleWindowEvent( eraseEvent );
+
+    wxNcPaintEvent ncPaintEvent( GetId() );
```
Comment (vadz) [inline src/qt/window.cpp]:
```suggestion
    wxNcPaintEvent ncPaintEvent( id );
```

File: src/qt/window.cpp [hunk 2]
```diff
@@ -393,13 +393,35 @@ void wxWindowQt::WarpPointer(int x, int y)
...
+    ncPaintEvent.SetEventObject( this );
+    HandleWindowEvent( ncPaintEvent );
+
+    wxPaintEvent paintEvent( id );
+    paintEvent.SetEventObject( this );
+    HandleWindowEvent( paintEvent );
+
```
Comment (vadz) [inline src/qt/window.cpp]:
```suggestion
```

File: src/qt/window.cpp [hunk 3]
```diff
@@ -393,13 +393,35 @@ void wxWindowQt::WarpPointer(int x, int y)
...
     if ( QtGetScrollBarsContainer() )
     {
-        QtGetScrollBarsContainer()->viewport()->update();
+        widget = QtGetScrollBarsContainer()->viewport();
     } else {
-        GetHandle()->update();
+        widget = GetHandle();
     }
+
+    widget->repaint();
```
Comment (vadz) [inline src/qt/window.cpp]: IUIC `repaint()` will actually force the window repaint immediately, so it seems strange to send the events _after_ the window had been already repainted.

In fact, I don't really understand how does this work: if `repaint()` updates the window synchronously, it must generate a paint event internally, how else would it know what to paint?

**Review by vadz — COMMENTED** [review]
I have some doubts about whether this is really correct. If you're sure it is and have tested that it does the right thing, I can still merge it, but please do test it carefully because it looks very suspicious.

I also wonder why does the PR mention `Refresh()` when it actually modifies `Update()`...

File: src/qt/window.cpp (hunk 3 above)
Comment (ffa-grahamdawes) [reply src/qt/window.cpp]: Sorry for the confusion over the pull request title I was intending to write Update not Refresh,  don't know if I can rename it somehow.

The call to `repaint()` doesn't result in any wx events being generated.  I'll take a look and see if I can understand why this is the case.

Comment (ffa-grahamdawes) [reply src/qt/window.cpp]: I was trying to get the EventPropagationTests passing for wxQT.  It looks like Update should be calling `repaint()` and not `update()` (which is more like wxWindow::Refresh) but you are correct QT will generate the correct events.

In the test case,  `QWidget::repaint()` is returning early as an isVisible check is failing.    Will submit a new pull request which just the repaint change then look at the failing test case separately,.

Comment (catalinr) [reply src/qt/window.cpp]: > I was trying to get the EventPropagationTests passing for wxQT.

Personally I always thought that tests were there to ensure that functionality [after being fixed] would not get broken [again]. Making a purpose of "passing tests", and applying changes that only do that does not imply that the functionality was fixed, and that IMO is worrying.

Comment (ffa-grahamdawes) [reply src/qt/window.cpp]: I think that generally depends on how good the test coverage is.  With good test coverage tests can serve as the specification to which an implementation must conform (this is basically the whole point of TDD or BDD).  

In other words a failing test means that the implementation doesn't match the specification. 

While I appreciate that there are many areas on wxWidgets that don't have coverage so getting the test to pass isn't the end of a particular fix,   I do find failing tests concerning as it implies a difference in behaviour between wxQT and wxGTK or wxMSW or any of the other ports. 

So to summarise I'm not not making changes *just* to pass the tests I do think it's important that the tests pass (if that makes sense).

**Comment by ffa-grahamdawes** [comment]
I'll close this request and try another approach.  I do think that the EventPropagationTest is highlighting a genuine problem in wxQT (with painting and event handling in general) but I'll need to some additional investigation.
